scalyr_agent/builtin_monitors/kubernetes_monitor.py

```diff
@@ -2983,16 +2985,114 @@ def __get_last_request_for_log(self, path):
+
+            # Check for the pod glob
+            pod_glob = config.get("k8s_pod_glob", None)
+            pod_name = k8s_info.get("pod_name", None)
+            if pod_glob is not None and pod_name is not None:
+                # ignore this config if pod_name doesn't match the glob
+                if not fnmatch.fnmatch(pod_name, pod_glob):
+                    self._logger.log(
+                        scalyr_logging.DEBUG_LEVEL_2,
+                        "Ignoring k8s_log item %d because pod_name '%s' doesn't match '%s'"
+                        % (i, pod_name, pod_glob,),
+                    )
+                    continue
+
+            # Check for the namespace glob
+            namespace_glob = config.get("k8s_namespace_glob", None)
+            pod_namespace = k8s_info.get("pod_namespace", None)
+            if namespace_glob is not None and pod_namespace is not None:
+                # ignore this config if pod_namespace doesn't match the glob
+                if not fnmatch.fnmatch(pod_namespace, namespace_glob):
+                    self._logger.log(
+                        scalyr_logging.DEBUG_LEVEL_2,
+                        "Ignoring k8s_log item %d because pod_namespace '%s' doesn't match '%s'"
+                        % (i, pod_namespace, namespace_glob,),
+                    )
+                    continue
+
+            # Check for the k8s container name glob
+            container_glob = config.get("k8s_container_glob", None)
+            k8s_container = k8s_info.get("k8s_container_name", None)
+            if container_glob is not None and k8s_container is not None:
+                # ignore this config if k8s_container doesn't match the glob
+                if not fnmatch.fnmatch(k8s_container, container_glob):
+                    self._logger.log(
+                        scalyr_logging.DEBUG_LEVEL_2,
+                        "Ignoring k8s_log item %d because k8s_container_name '%s' doesn't match '%s'"
+                        % (i, k8s_container, container_glob,),
+                    )
+                    continue
+
+            self._logger.log(
+                scalyr_logging.DEBUG_LEVEL_2,
+                "Applying k8s_log config item %d.  Matched pod_name ('%s', '%s'), pod_namespace ('%s', '%s') and k8s_container_name ('%s', '%s')"
+                % (
+                    i,
+                    pod_name,
+                    pod_glob,
+                    pod_namespace,
+                    namespace_glob,
+                    k8s_container,
+                    container_glob,
+                ),
+            )
+            # We have the first matching config.  Apply the log config and break
+            # Note, we can't just .update() because otherwise the attributes dict
+            # may get overridden, plus we also need to exclude `path`
+            for key, value in six.iteritems(config):
+                # Ignore `path` so people can't override it
```

Comment:
Hmm.. so you are essentially performing a merge with the default attributes on every match you get.  When you move this to an abstraction, I would advocate you merge in the default attributes during the instance creation so that you don't have to do this over and over again.

Reply:
After further thought, we need to be doing this, because it's not the default attributes we are merging, but the container specific attributes (pod_name, pod_namespace etc).  We need to do that for every container anyway so I figure it's better to do it once rather than holding state.